src/View/Window.vala

```diff
@@ -323,78 +351,150 @@ public class Files.View.Window : Hdy.ApplicationWindow {
-            add_tab_by_uri (restore_data);
-        });
+        if (!path_in_menu) {
+            menu.append (
+                FileUtils.sanitize_path (path, null, false),
+                "win.tabhistory-restore::%s".printf (path)
+            );
+        }
 
-        tabs.tab_duplicated.connect ((tab) => {
-            add_tab_by_uri (((ViewContainer)(tab.page)).uri);
-        });
+        view_container.close ();
+        tab_view.close_page_finish (page, true);
+
+        if (tab_view.n_pages == 0) {
+            add_tab ();
+        }
 
-        tabs.tab_moved.connect ((tab) => {
-            /* Called when tab dragged out of notebook */
-            var vc = (ViewContainer)(tab.page) ;
-            /* Close view now to disconnect signal handler closures which can trigger after slot destruction */
-            vc.close ();
+        return Gdk.EVENT_STOP;
+    }
 
-            marlin_app.create_window (vc.location, real_mode (vc.view_mode));
+    private void tab_view_setup_menu (Hdy.TabPage? page) {
+        if (page == null) {
+            return;
+        }
 
-            /* remove_tab function uses Idle loop to close tab */
-            remove_tab (tab);
-        });
+        var action_close = new SimpleAction ("tabmenu-close", null);
+        var action_close_end = new SimpleAction ("tabmenu-close-end", null);
+        var action_close_others = new SimpleAction ("tabmenu-close-others", null);
+        var action_duplicate = new SimpleAction ("tabmenu-duplicate", null);
+        var action_new_window = new SimpleAction ("tabmenu-new-window", null);
 
+        add_action (action_close);
+        add_action (action_close_end);
+        add_action (action_close_others);
+        add_action (action_duplicate);
+        add_action (action_new_window);
 
-        tabs.tab_added.connect ((tab) => {
-            var vc = (ViewContainer)(tab.page) ;
-            vc.window = this;
-        });
+        marlin_app.set_accels_for_action ("win.tabmenu-close", {"<Ctrl>W"});
+        marlin_app.set_accels_for_action ("win.tabmenu-duplicate", {"<Ctrl><Alt>T"});
+        marlin_app.set_accels_for_action ("win.tabmenu-new-window", {"<Ctrl><Alt>N"});
 
-        tabs.tab_removed.connect (on_tab_removed);
+        var tab_menu = (Menu) tab_view.menu_model;
+        tab_menu.remove_all ();
 
-        sidebar.request_focus.connect (() => {
-            return !current_tab.locked_focus && !locked_focus;
+        var open_tab_section = new Menu ();
+        open_tab_section.append (_("Open in New Window"), "win.tabmenu-new-window");
+        open_tab_section.append (_("Duplicate Tab"), "win.tabmenu-duplicate");
+
+        var close_tab_section = new Menu ();
+        close_tab_section.append (_("Close Tabs to the Right"), "win.tabmenu-close-end");
+        close_tab_section.append (_("Close Other Tabs"), "win.tabmenu-close-others");
+        /// TRANSLATORS: For RTL this should be "to the left"
```

Comment:
Misplaced translators comment - refers to to lines above.

Reply:
Oops thanks for catching that